File: DSA_Codesheet/Java/Merge_Sort.java

import java.util.Scanner;

public class Merge_Sort {
    public static void main(String[] args) {
        Scanner s = new Scanner(System.in);
        System.out.println("Enter size of the list:");
        int n = s.nextInt();
        int[] arr = new int[n];
        System.out.println("Enter the elements of the list:");
        for (int i = 0; i < n; i++) {
            arr[i] = s.nextInt();
        }
        s.close();
        int low = 0;
        int high = n-1;
        mergeSort(arr,low,high);
        for (int i:arr) {
            System.out.print(i+" ");
        }

    }
    static void mergeSort(int[] arr,int low,int high){
        if(low<high){  //to check if size of array is greater than 1
            int mid = low+(high-low)/2;
            mergeSort(arr,low,mid);
            mergeSort(arr,mid+1,high);
            merge(arr,low,mid,high);
        }
    }
    static void merge(int[] arr,int low,int mid,int high){
        int n1 = mid-low+1;   //calculating size of left array
        int n2 = high-mid;    //size of right array
        int[] left = new int[n1];
        int[] right = new int[n2];
        for (int i = 0; i < n1; i++) {
            left[i] = arr[low+i];
        }
        for (int i = 0; i < n2; i++) {
            right[i] = arr[mid+i+1];
        }
        int i = 0,j=0,k=low;
        while (i<n1 && j<n2){
            if(left[i]<=right[j]){
                arr[k] = left[i];
                i++;
                k++;
            }else {
                arr[k] = right[j];
                j++;
                k++;
            }
        }
        while (i<n1){
            arr[k] = left[i];
            i++;
            k++;
        }
        while (j<n2){
            arr[k] = right[j];
            j++;
            k++;
        }
    }
}

File: DSA_Codesheet/Python/merge_sort.py
def merge(arr, l, m, r):

  n1 = m - l + 1
  n2 = r - m
  L = [0] * n1
  R = [0] * n2

  for i in range(0, n1):
    L[i] = arr[l + i]

  for j in range(0, n2):
    R[j] = arr[m + 1 + j]

  i = 0
  j = 0
  k = l
  while i < n1 and j < n2:
    if L[i] <= R[j]:
      arr[k] = L[i]
      i += 1
    else:
      arr[k] = R[j]
      j += 1
    k += 1

  while i < n1:
    arr[k] = L[i]
    i += 1
    k += 1

  while j < n2:
    arr[k] = R[j]
    j += 1
    k += 1


def merge_sort(arr, l, r):

  if l < r:
    m = (l + r) // 2
    merge_sort(arr, l, m)
    merge_sort(arr, m + 1, r)
    merge(arr, l, m, r)


arr = input("Enter a list of numbers: ").split(" ")

arr = [int(i) for i in arr]

merge_sort(arr, 0, len(arr) - 1)

for i in arr:
  print(i,end=' ')
